src/rust/catloop/mod.rs

```diff
@@ -428,13 +446,17 @@ impl CatloopLibOS {
                 Some(duplex_pipe) => duplex_pipe.rx(),
                 None => unreachable!("pop() an unconnected queue"),
             },
-            None => return Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
+            None => {
+                let cause: String = format!("invalid queue descriptor (qd={:?})", qd);
+                error!("pop(): {:?}", &cause);
+                return Err(Fail::new(libc::EBADF, &cause));
+            },
         };
 
         let qt: QToken = self.catmem.borrow_mut().pop(catmem_qd, size)?;
-        self.qts.insert(qt, (demi_opcode_t::DEMI_OPC_POP, qd));
+        self.catmem_qts.insert(qt, (demi_opcode_t::DEMI_OPC_POP, qd));
 
-        Ok(qt)
+        Ok(Self::shift_qtoken(qt))
```

Comment:
Not super important, but you could probably get away with a single HashMap if you use the shifted qtoken.

Reply:
Are you suggesting that we should replace this by a hashmap? We already have one, it is important to actually distribute unique queue tokens to applications.

Reply:
No, I'm saying if you use the shifted value, you can keep both sets of qtokens in the same hash map because the range no longer overlaps. Again, this is not super important.

Reply:
Oh, cool idea! I will work on this.

Reply:
I'll take this optimization in a different PR. I'm submitting an issue for this.